pkg/builtin/unary/reverse.go

```diff
@@ -0,0 +1,92 @@
+// Copyright 2021 Matrix Origin
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package unary
+
+import (
+	"fmt"
+
+	"github.com/matrixorigin/matrixone/pkg/builtin"
+	"github.com/matrixorigin/matrixone/pkg/container/nulls"
+	"github.com/matrixorigin/matrixone/pkg/container/types"
+	"github.com/matrixorigin/matrixone/pkg/container/vector"
+	"github.com/matrixorigin/matrixone/pkg/sql/colexec/extend"
+	"github.com/matrixorigin/matrixone/pkg/sql/colexec/extend/overload"
+	"github.com/matrixorigin/matrixone/pkg/vectorize/reverse"
+	"github.com/matrixorigin/matrixone/pkg/vm/process"
+)
+
+var ReverseArgAndRets = []argsAndRet{
+	{[]types.T{types.T_char}, types.T_char},
+	{[]types.T{types.T_varchar}, types.T_varchar},
+}
+
+func init() {
+	// register built-in func and its args and rets
+	extend.FunctionRegistry["reverse"] = builtin.Reverse
+	overload.OpTypes[builtin.Reverse] = overload.Unary
+
+	for _, item := range ReverseArgAndRets {
+		overload.AppendFunctionRets(builtin.Reverse, item.args, item.ret)
+	}
+
+	extend.UnaryReturnTypes[builtin.Reverse] = func(extend extend.Extend) types.T {
+		return getUnaryReturnType(builtin.Reverse, extend)
+	}
+	// add built-in stringify
+	extend.UnaryStrings[builtin.Reverse] = func(extend extend.Extend) string {
+		return fmt.Sprintf("reverse(%s)", extend)
+	}
+
+	overload.UnaryOps[builtin.Reverse] = []*overload.UnaryOp{
+		{
+			Typ:        types.T_varchar,
+			ReturnType: types.T_varchar,
+			Fn: func(inputVec *vector.Vector, proc *process.Process, b bool) (*vector.Vector, error) {
+				inputVecCol := inputVec.Col.(*types.Bytes)
+				resultVec, err := process.Get(proc, int64(len(inputVecCol.Data)), types.Type{Oid: types.T_varchar, Size: 24})
```

Comment:
I think we could reuse the original vector when possible

Reply:
I get it, but still has a question to ask.
I notice those code in the tutorial
```go
if origVec.Ref == 1 || origVec.Ref == 0 { // reuse the original vector when we don't need the original one anymore
					origVec.Ref = 0
					abs.AbsFloat32(origVecCol, origVecCol)
					return origVec, nil
				}
```
I want to know in which situation make `ref == 0`, `ref == 1`, `ref > 1` or what do they mean seprately?

From my point, `ref == 1` means those column data stored in the vector is only occupied by this process, and `ref == 0` means they are not occupied by any process. In this situation, they are free to change.

Reply:
this reference count mechanism has some flaws and we may change it later, for now, we could think of it as follows.
suppose we have a query like this
```sql
select reverse(a) from table1;
```
before the computation begins, we compile this query to a computation plan, and we use this ref variable to mark that this vector a is used only here, therefore we can change it during our computation.

but suppose we have another query:
```sql
select reverse(a), a from table1;
```
then when we compute the reverse result of a, we can't change the original value of a since we need it later, reference count is the solution now for this kind of scenario

Reply:
Ref wasn't the number of processes which use the vector's memory. Each process has its own memory pool.

Just consider these SQLs
1. "select a, a+1 from t"

in this case, vector a (from the storage engine)'s ref is 2, because it occurs in a and a+1, and we will do projection work for them.
when we calculate the a+1, we found that, ref of a is 2, and we can't reuse its memory.

2. "select a+1 from t" 

in this case when we calculate the a+1, ref of a is 1 (read from storage). and a has the same type with a+1, so we will reuse its memory.

And Ref 0 is special case. When 0, it means these memory can be free.

Reply:
Thx for your answer. I kind of understand what the `ref` means;